feat: Fast Bridge unhappy path

jaybuidl opened the pull request:
Resolves #53

shotaronowhere commented:
Should use `storage` instead of `memory` to persist state changes in struct L94, L134, etc.

I [fixed](https://github.com/kleros/kleros-v2/blob/feat/challenge-path-2-with-merkle-proof/contracts/src/bridge/FastBridgeSenderToEthereum.sol) the `registered` mapping logic in the challenge-path-2-merkle-root branch. I think the single message fast bridge should implement a similar pattern for safe send.

Apologies, the fix I just pushed was supposed to be included in the original push, but I had 2 code editors open at the same time and on overwrote the other.

jaybuidl commented:
Why close the PR thou..

jaybuidl commented:
> Should use `storage` instead of `memory` to persist state changes in struct L94, L134, etc.
Fixed

shotaronowhere commented on contracts/src/bridge/FastBridgeSenderToEthereum.sol:
```diff
@@ -0,0 +1,145 @@
+        bytes memory _calldata
+    ) external payable override {
+        Ticket storage ticket = tickets[_ticketID];
+        require(ticket.messageHash != 0, "Ticket does not exist.");
```
Alternatively a null pattern can be used in the fastBridgeSender so that honest challengers can prove ticketIDs do not exist. Otherwise, challenges to claims on the fastBridgeReceiver about ticketIDs that do not exist cannot be resolved.

Moreover there is a *zero cost grief where the challenger and claimant are the same. A dishonest claimant can make a claim about a ticketID that does not exist, and the same dishonest actor can immediately challenge the claim. When the claim is eventually resolved, the challenger who is the same as the claimant can claim their own deposit back. Maybe 50% of the deposit should be burned or set towards rewarding future relayers / bridgers to ensure an immediate dishonest claim and challenge does not result in a zero cost delay grief.

jaybuidl commented on contracts/src/bridge/FastBridgeReceiverOnEthereum.sol:
```diff
@@ -0,0 +1,202 @@
+
+        // Claim assessment if any
+        bytes32 messageHash = keccak256(abi.encode(_ticketID, _messageData));
+        if (ticket.claim.bridger != address(0) && ticket.claim.messageHash == messageHash) {
```
Good find! Fixed with `require(ticket.relayed == false)` inside `claim()`

jaybuidl commented:
> 2. Unpredictable hashes using the block numbers.

Implemented

> 3. Challenger reward lower than the Bridger's one.

Can be done via configuration/governance.

> 4. Claims rate-limited to a period equivalent to the batching period.

Will be done in the branch [feat/challenge-path-with-merkle-proof](https://github.com/kleros/kleros-v2/tree/feat/challenge-path-with-merkle-proof).

> 1. Null message to prove the non-existence of a ticketID.

Implemented this in a separate branch [feat/challenge-path-2-null-message](https://github.com/kleros/kleros-v2/commit/abb86876f7a5c5da113b1479fb9866b9f43a3777), it introduces extra complexity and edge cases which may not be necessary given 2. 3. and 4. will be done.

shotaronowhere commented:
> > 3. Challenger reward lower than the Bridger's one.
> 
> Can be done via configuration/governance.

Governance / configuration can only change the challenger and claim deposit size. In order to change the pay-out of the challenger, we need to modify [L170](https://github.com/kleros/kleros-v2/blob/7bc2272bac0153eeff962cbdaeda59aa970c1c28/contracts/src/bridge/FastBridgeReceiverOnEthereum.sol#L170)
To something like:

'    uint public constant ALPHA_DIVISOR = 1e4;'
'    uint public rewardBurnRatio = 5e3; // half reward, half burnt'



`        uint256 amount = ticket.claim.claimDeposit*rewardBurnRatio/ALPHA_DIVISOR + ticket.challenge.challengeDeposit;`

And let governance update / change the rewardBurnRatio. Or to be honest, maybe rewardBurnRatio should be immutable. 

To make sure that a dishonest claim -> self challenge does not result in a zero-cost delay grief.

hrishibhat reviewed (dismissed):
Looks good.
